File: agentcash/crypto.py
"""Proof of possession.

v0.1 bound tokens with a shared secret the ledger stored. That worked, but it
made the ledger a trusted hub and - worse - it meant a delegation could not be
completed without calling home to register the child's secret.

Ed25519 fixes both. The public key travels *inside the token*, inside the
signature chain, so the ledger needs to store nothing and delegation is fully
offline: an agent on a plane, network down, can still carve a bounded wallet
for a sub-agent and that wallet will verify the first time anyone sees it.

The HMAC backend is kept as a fallback for environments without `cryptography`.
It is honestly labelled as hub-trust in `Ledger.security_profile()`; nothing
silently downgrades.
"""

import hashlib
import hmac
import secrets as _secrets

try:
    from cryptography.hazmat.primitives.asymmetric.ed25519 import (
        Ed25519PrivateKey, Ed25519PublicKey,
    )
    from cryptography.exceptions import InvalidSignature
    HAVE_ED25519 = True
except Exception:                                    # pragma: no cover
    HAVE_ED25519 = False

ED25519 = "ed25519"
HMAC_SHA256 = "hmac-sha256"


class Signer:
    """What the agent holds. Never leaves the agent's process."""

    alg = None

    def public(self) -> str:
        raise NotImplementedError

    def sign(self, msg: str) -> str:
        raise NotImplementedError

    def binding(self) -> dict:
        """The `cnf` caveat value - travels inside the token, publicly."""
        return {"alg": self.alg, "pub": self.public()}

    def ledger_secret(self):
        """What (if anything) the ledger must store to verify us.

        Ed25519: None. That is the entire point.
        """
        return None

    def to_dict(self):
        raise NotImplementedError

    @staticmethod
    def from_dict(d):
        if d["alg"] == ED25519:
            return Ed25519Signer(bytes.fromhex(d["sk"]))
        return HmacSigner(d["sk"])


class Ed25519Signer(Signer):
    alg = ED25519

    def __init__(self, sk_bytes=None):
        if not HAVE_ED25519:                          # pragma: no cover
            raise RuntimeError("ed25519 backend unavailable")
        self._sk = (Ed25519PrivateKey.from_private_bytes(sk_bytes)
                    if sk_bytes else Ed25519PrivateKey.generate())

    def public(self) -> str:
        from cryptography.hazmat.primitives import serialization
        return self._sk.public_key().public_bytes(
            serialization.Encoding.Raw, serialization.PublicFormat.Raw
        ).hex()

    def sign(self, msg: str) -> str:
        return self._sk.sign(msg.encode()).hex()

    def to_dict(self):
        from cryptography.hazmat.primitives import serialization
        raw = self._sk.private_bytes(
            serialization.Encoding.Raw, serialization.PrivateFormat.Raw,
            serialization.NoEncryption(),
        )
        return {"alg": self.alg, "sk": raw.hex()}


class HmacSigner(Signer):
    alg = HMAC_SHA256

    def __init__(self, secret=None):
        self._secret = secret or _secrets.token_hex(32)

    def public(self) -> str:
        return hashlib.sha256(self._secret.encode()).hexdigest()

    def sign(self, msg: str) -> str:
        return hmac.new(self._secret.encode(), msg.encode(), hashlib.sha256).hexdigest()

    def ledger_secret(self):
        return self._secret

    def to_dict(self):
        return {"alg": self.alg, "sk": self._secret}


def generate(prefer_ed25519=True) -> Signer:
    if prefer_ed25519 and HAVE_ED25519:
        return Ed25519Signer()
    return HmacSigner()


def verify(binding: dict, msg: str, sig: str, ledger_secret=None) -> bool:
    """Verify a proof against the binding carried in the token."""
    alg, pub = binding.get("alg"), binding.get("pub")
    if not sig:
        return False
    if alg == ED25519:
        if not HAVE_ED25519:                          # pragma: no cover
            return False
        try:
            Ed25519PublicKey.from_public_bytes(bytes.fromhex(pub)).verify(
                bytes.fromhex(sig), msg.encode()
            )
            return True
        except (InvalidSignature, ValueError):
            return False
    if alg == HMAC_SHA256:
        if ledger_secret is None:
            return False
        if hashlib.sha256(ledger_secret.encode()).hexdigest() != pub:
            return False
        expect = hmac.new(ledger_secret.encode(), msg.encode(), hashlib.sha256).hexdigest()
        return hmac.compare_digest(expect, sig)
    return False


def profile() -> dict:
    return {
        "default_alg": ED25519 if HAVE_ED25519 else HMAC_SHA256,
        "ed25519_available": HAVE_ED25519,
        "ledger_stores_private_material": not HAVE_ED25519,
        "offline_delegation": HAVE_ED25519,
    }

File: js/src/crypto.js
// Ed25519 proof of possession, on node:crypto only. No dependencies.
//
// The public key travels inside the token's signature chain, so a verifier
// needs no prior knowledge of a delegated wallet and delegation works offline.

import { createHmac, createHash, createPrivateKey, createPublicKey,
         sign as nodeSign, verify as nodeVerify, generateKeyPairSync,
         randomBytes, timingSafeEqual } from "node:crypto";

export const ED25519 = "ed25519", HMAC_SHA256 = "hmac-sha256";

// RFC 8410 DER wrappers, so raw 32-byte keys round-trip through node:crypto.
const SPKI  = Buffer.from("302a300506032b6570032100", "hex");
const PKCS8 = Buffer.from("302e020100300506032b657004220420", "hex");
const pubFromRaw  = raw => createPublicKey({
  key: Buffer.concat([SPKI, raw]), format: "der", type: "spki" });
const privFromRaw = raw => createPrivateKey({
  key: Buffer.concat([PKCS8, raw]), format: "der", type: "pkcs8" });

export class Ed25519Signer {
  static alg = ED25519;
  constructor(rawSk) {
    if (!rawSk) {
      const { privateKey } = generateKeyPairSync("ed25519");
      rawSk = privateKey.export({ format: "der", type: "pkcs8" }).subarray(PKCS8.length);
    }
    this.alg = ED25519;
    this._raw = Buffer.from(rawSk);
    this._sk = privFromRaw(this._raw);
  }
  public() {
    return this._sk.export({ format: "jwk" }).x
      ? Buffer.from(this._sk.export({ format: "jwk" }).x, "base64url").toString("hex")
      : null;
  }
  sign(msg) { return nodeSign(null, Buffer.from(msg, "utf8"), this._sk).toString("hex"); }
  binding() { return { alg: this.alg, pub: this.public() }; }
  ledgerSecret() { return null; }
  toJSON() { return { alg: this.alg, sk: this._raw.toString("hex") }; }
}

export class HmacSigner {
  constructor(secret) {
    this.alg = HMAC_SHA256;
    this._secret = secret || randomBytes(32).toString("hex");
  }
  public() { return createHash("sha256").update(this._secret).digest("hex"); }
  sign(msg) { return createHmac("sha256", this._secret).update(msg, "utf8").digest("hex"); }
  binding() { return { alg: this.alg, pub: this.public() }; }
  ledgerSecret() { return this._secret; }
  toJSON() { return { alg: this.alg, sk: this._secret }; }
}

export const signerFromJSON = d =>
  d.alg === ED25519 ? new Ed25519Signer(Buffer.from(d.sk, "hex")) : new HmacSigner(d.sk);

export const generate = (preferEd = true) =>
  preferEd ? new Ed25519Signer() : new HmacSigner();

export function verify(binding, msg, sig, ledgerSecret = null) {
  if (!sig) return false;
  try {
    if (binding.alg === ED25519) {
      return nodeVerify(null, Buffer.from(msg, "utf8"),
        pubFromRaw(Buffer.from(binding.pub, "hex")), Buffer.from(sig, "hex"));
    }
    if (binding.alg === HMAC_SHA256) {
      if (ledgerSecret === null) return false;
      if (createHash("sha256").update(ledgerSecret).digest("hex") !== binding.pub) return false;
      const a = Buffer.from(createHmac("sha256", ledgerSecret).update(msg, "utf8").digest("hex"));
      const b = Buffer.from(sig);
      return a.length === b.length && timingSafeEqual(a, b);
    }
  } catch { return false; }
  return false;
}

export const profile = () => ({
  default_alg: ED25519, ed25519_available: true,
  ledger_stores_private_material: false, offline_delegation: true,
});
